platform/linux-generic/pktio/socket_xdp.c

```diff
@@ -0,0 +1,676 @@
+#include <odp_socket_common.h>
+#include <odp_config_internal.h>
+
+#include <string.h>
+#include <errno.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#include <poll.h>
+
+#include <xdp/xsk.h>
+
+#define NUM_XDP_DESCS 1024U
+#define MIN_FRAME_SIZE 2048U
+#define IF_DELIM " "
+#define Q_DELIM ':'
+
+typedef struct {
+	struct xsk_ring_prod fill_q;
+	struct xsk_ring_cons compl_q;
+	struct xsk_umem *umem;
+	pool_t *pool;
+} xdp_umem_info_t;
+
+typedef struct {
+	struct xsk_ring_cons rx;
+	struct xsk_ring_cons compl_q;
+	struct xsk_ring_prod tx;
+	struct xsk_ring_prod fill_q;
+	xdp_umem_info_t *umem_info;
+	struct xsk_socket *xsk;
+	int pktio_idx;
+	int helper_sock;
+	uint32_t mtu;
+	uint32_t max_mtu;
+} xdp_sock_info_t;
+
+typedef struct {
+	odp_ticketlock_t rx_lock ODP_ALIGNED_CACHE;
+	odp_ticketlock_t tx_lock ODP_ALIGNED_CACHE;
+	xdp_sock_info_t sock_info;
+} pkt_xdp_t;
+
+typedef struct {
+	odp_packet_hdr_t *odp_hdr;
+	odp_packet_t odp_pkt;
+	uint8_t *pkt;
+	uint32_t len;
+} pkt_data_t;
+
+ODP_STATIC_ASSERT(PKTIO_PRIVATE_SIZE >= sizeof(pkt_xdp_t),
+		  "PKTIO_PRIVATE_SIZE too small");
+
+static odp_bool_t disable_pktio;
+
+static int sock_xdp_init_global(void)
+{
+	if (getenv("ODP_PKTIO_DISABLE_SOCKET_XDP")) {
+		ODP_PRINT("PKTIO: socket xdp skipped,"
+			  " enabled export ODP_PKTIO_DISABLE_SOCKET_XDP=1.\n");
+		disable_pktio = true;
+	} else {
+		ODP_PRINT("PKTIO: initialized socket xdp,"
+			  " use export ODP_PKTIO_DISABLE_SOCKET_XDP=1 to disable.\n");
+	}
+
+	return 0;
+}
+
+static inline pkt_xdp_t *pkt_priv(pktio_entry_t *pktio_entry)
+{
+	return (pkt_xdp_t *)(uintptr_t)(pktio_entry->s.pkt_priv);
+}
+
+static uint32_t get_bind_queue_index(const char *devname)
+{
+	const char *param = getenv("ODP_PKTIO_XDP_PARAMS");
+	char *tmp_str;
+	char *tmp;
+	char *if_str;
+	int idx = 0;
+
+	if (param == NULL)
+		goto out;
+
+	tmp_str = strdup(param);
+
+	if (tmp_str == NULL)
+		goto out;
+
+	tmp = strtok(tmp_str, IF_DELIM);
+
+	if (tmp == NULL)
+		goto out_str;
+
+	while (tmp) {
+		if_str = strchr(tmp, Q_DELIM);
+
+		if (if_str != NULL && if_str != &tmp[strlen(tmp) - 1U]) {
+			if (strncmp(devname, tmp, (uint64_t)(uintptr_t)(if_str - tmp)) == 0) {
+				idx = MAX(atoi(++if_str), 0);
+				break;
+			}
+		}
+
+		tmp = strtok(NULL, IF_DELIM);
+	}
+
+out_str:
+	free(tmp_str);
+
+out:
+	return idx;
+}
+
+static odp_bool_t reserve_fill_queue_elements(xdp_sock_info_t *sock_info, int num)
+{
+	pool_t *pool;
+	odp_packet_t packets[num];
+	int count;
+	struct xsk_ring_prod *fill_q;
+	uint32_t start_idx;
+	int pktio_idx;
+	odp_packet_hdr_t *pkt_hdr;
+
+	pool = sock_info->umem_info->pool;
+	count = odp_packet_alloc_multi(pool->pool_hdl, sock_info->mtu, packets, num);
+
+	if (count <= 0) {
+		ODP_ERR("Unable to allocate packets for fill queue.\n");
+		return false;
+	}
+
+	fill_q = &sock_info->fill_q;
+
+	if (xsk_ring_prod__reserve(fill_q, count, &start_idx) == 0U) {
+		ODP_ERR("Unable to reserve fill queue descriptors.\n");
```

Comment:
If pool becomes empty in the fast path this will flood the log. Likely same issue for the `xsk_ring_prod__reserve()` below .

Reply:
Moved the error to be logged in `sock_xdp_open()`, will add suitable error counters later to keep track of fast path errors.